File: 0053-maximum-subarray/0053-maximum-subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        # Kadane's Algorithm
        cs = 0
        ms = float(-inf)

        for n in nums:
            if cs < 0:
                cs = 0
            cs += n
            ms = max(cs, ms)
        return ms


# cs = CurrentSum | ms = MaxSum
# Time Complexity : O(n)
# Space Complexity : O(1)
# by ar - sayeem[January 20, 2026]

File: 0053-maximum-subarray/0053-maximum-subarray.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        // Kadane's Algorithm
        int CS = 0, MS = INT_MIN;
        for(int i : nums)
        {
            CS += i;
            MS = max(CS, MS);
            if(CS < 0)
            {
                CS = 0;
            }
        }

        return MS;
    }
};

// CS = CurrentSum | MS = MaxSum
// Time Complexity : O(n)
// Space Complexity : O(1)
// by ar - sayeem[Month Date, 2026]